### ISSUE TITLE
APIError: Unknown requirement type 'DISKSPACE' or argument 5000

### ISSUE THREAD

**Issue by thomas-mueller** [grid-control/grid-control#45]
Dear experts,

I updated to tag r1982 and observe the following problem
```
BackendError: Could not write jdl data to /tmp/tmpVFZVSU.jdl.
----------
APIError: Unknown requirement type 'DISKSPACE' or argument 5000
```
which seems to be caused by [this monster commit](https://github.com/grid-control/grid-control/commit/c2d83cbe6da6ae2f825c49396cbb086c04242073). I do not specify `scratch space used`, but if I do so, I get the same error with modified argument.

The following changes seem to work as a temporary solution in my case:
```
diff --git a/packages/grid_control/backends/wms.py b/packages/grid_control/backends/wms.py
index 38c5b08..aa4d780 100644
--- a/packages/grid_control/backends/wms.py
+++ b/packages/grid_control/backends/wms.py
@@ -111,7 +111,7 @@ class WMS(NamedPlugin):
                return tuple(gc_id.split('.', 2)[1:])
 
 make_enum(['WALLTIME', 'CPUTIME', 'MEMORY', 'CPUS', 'BACKEND',
-       'SITES', 'QUEUES', 'SOFTWARE', 'STORAGE', 'DISKSPACE'], WMS)
+       'SITES', 'QUEUES', 'SOFTWARE', 'STORAGE'], WMS)
 
 
 class BasicWMS(WMS):
diff --git a/packages/grid_control/parameters/psource_basic.py b/packages/grid_control/parameters/psource_basic.py
index 9d1ac07..ea63e5c 100644
--- a/packages/grid_control/parameters/psource_basic.py
+++ b/packages/grid_control/parameters/psource_basic.py
@@ -56,7 +56,7 @@ class RequirementParameterSource(ParameterSource):
                        result[ParameterInfo.REQS].append((WMS.WALLTIME, parse_time(result.pop('WALLTIME'))))
                if 'CPUTIME' in result:
                        result[ParameterInfo.REQS].append((WMS.CPUTIME, parse_time(result.pop('CPUTIME'))))
-               for (req_key, vn) in [(WMS.MEMORY, 'MEMORY'), (WMS.CPUS, 'CPUS'), (WMS.DISKSPACE, 'DISKSPACE')]:
+               for (req_key, vn) in [(WMS.MEMORY, 'MEMORY'), (WMS.CPUS, 'CPUS')]:
                        if vn in result:
                                result[ParameterInfo.REQS].append((req_key, int(result.pop(vn))))
 
diff --git a/packages/grid_control/tasks/task_base.py b/packages/grid_control/tasks/task_base.py
index f8ed859..826f4d2 100644
--- a/packages/grid_control/tasks/task_base.py
+++ b/packages/grid_control/tasks/task_base.py
@@ -147,7 +147,6 @@ class TaskModule(NamedPlugin):  # pylint:disable=too-many-instance-attributes
                        (WMS.CPUTIME, self._cpu_time),
                        (WMS.MEMORY, self._memory),
                        (WMS.CPUS, self._cpu_min),
-                       (WMS.DISKSPACE, self._disk_min),
                ] + self._source.get_job_content(jobnum)[ParameterInfo.REQS]
 
        def get_sb_in_fpi_list(self):  # Get file path infos for input sandbox
```

**Comment by FredStober** [comment]
Ah, yes - the final code review commit which dropped the pylint / pep8 / flake8 message count to 0 :)

Since fixing some issues (in particular the code complexity warnings) necessitated refactoring / changes to the backend-task interface, I was quite concerned about this commit as well.

Unfortunately, I only got one reply to my email about this topic - though I admit the subject line was not very descriptive.

Anyway - the proper fix is:

```
diff --git a/packages/grid_control/backends/jdl_writer.py b/packages/grid_control/backends/jdl_writer.py
index 09eacdc7..7f70fc5c 100644
--- a/packages/grid_control/backends/jdl_writer.py
+++ b/packages/grid_control/backends/jdl_writer.py
@@ -62,6 +62,8 @@ class JDLWriter(object):
                                req_string_list.append(self._format_reqs_sites(arg))
                        elif req_type == WMS.CPUS:
                                pass  # Handle number of cpus in prepare
+                       elif req_type == WMS.DISKSPACE:
+                               pass  # no GlueCE attribute available
                        else:
                                raise APIError('Unknown requirement type %r or argument %r' % (WMS.enum2str(req_type), arg))
                result['Requirements'] = str.join(' && ', iidfilter(req_string_list))

```

If the CREAMCE **does** support specifying disk space requirements, subclassing `JDLWriter`, overwriting `_format_reqs` and adding: `GridWMS.__init__(self, ..., jdl_writer=NewJDLWriter())` to the init call of CreamWMS would allow to support this new requirement type.

**Comment by thomas-mueller** [comment]
Do I understand it right, that the monster commit plus your "proper fix" removes the proper setting of the scratch space requirement for the CREAMCE backend (which was available before), such that this setting is basically ignored?

**Comment by FredStober** [comment]
In the past, the scratch space was only checked while the job was running.

Since some batch systems support giving the disk space requirement already during submission time, the new WMS.DISKSPACE requirement was added (which is set to the same number as the runtime check). As far as I know, there is no GlueCE attribute for this - so this requirement should be ignored.

However the runtime check is still being run... so for CREAMCE nothing  changes.
The issue is just that the grid backend code is very cautious about unknown requirements.

**Comment by FredStober** [comment]
FYI:
Runtime variables concerning disk space monitoring are set here:
https://github.com/grid-control/grid-control/blob/master/packages/grid_control/tasks/task_base.py#L75
and used here:
https://github.com/grid-control/grid-control/blob/master/packages/grid_control/share/gc-run.sh#L66

**Comment by thomas-mueller** [comment]
Thanks for your explanations. Would you be so kind and add these two lines? Forking, changing, making a pull request, reakting on feedback and waiting for a merge seems to be a bit too much overhead for just two lines.

**Comment by FredStober** [comment]
Does everything work for you with just these two lines? (without the changes proposed in the first message)

**Comment by thomas-mueller** [comment]
Sorry, I did not submit new jobs after my first "fix". So it is not yet tested.

**Comment by thomas-mueller** [comment]
Now I tested it and the 2 lines solve the problem.

**Comment by FredStober** [comment]
Due to vacations, work related travels and lots of other stuff, I forgot to merge the fix and close this issue.

Fixed in e87d001223abf6b4be92c8c12ac3eae4dff7b519
